src/host/linalg.jl

```diff
@@ -197,33 +197,43 @@ LinearAlgebra.lmul!(a::Number, B::AbstractGPUArray) = generic_lmul!(a, B)
 LinearAlgebra.permutedims!(dest::AbstractGPUArray, src::AbstractGPUArray, perm) =
     permutedims!(dest, src, Tuple(perm))
 
+@inline @generated function permute_linearindex(size::NTuple{N,T}, l::T, strides::NTuple{N,T}) where {N,T}
+    quote
+        l -= one(T)
+        res = one(T)
+        Base.Cartesian.@nexprs $(N-1) i->begin
+            assume(size[i] > 0)
+            @inbounds l, s = divrem(l, size[i])
+            @inbounds res += s * strides[i]
+        end
+        return @inbounds res + strides[N] * l
+    end
+end
 function LinearAlgebra.permutedims!(dest::AbstractGPUArray, src::AbstractGPUArray,
                                     perm::NTuple{N}) where N
-    Base.checkdims_perm(dest, src, perm)
+    length(dest) <= typemax(UInt32) ? _permutedims!(UInt32, dest, src, perm) : _permutedims!(UInt64, dest, src, perm)
+end
 
-    # get the new strides of destination tensor
+function _permutedims!(::Type{IT}, dest::AbstractGPUArray, src::AbstractGPUArray,
+                                    perm::NTuple{N}) where {IT,N}
+    @assert length(src) <= typemax(IT)
+    Base.checkdims_perm(dest, src, perm)
     dest_strides = ntuple(k->k==1 ? 1 : prod(i->size(dest, i), 1:k-1), N)
-    dest_strides_perm = ntuple(i->dest_strides[findfirst(==(i), perm)], N)
-
-    function permutedims_kernel(ctx, dest, src, dest_strides_perm)
-        # find the cartesian index in source tensor
-        LI = @linearidx src
-        I = @inbounds CartesianIndices(src)[LI]
-
-        # the corresponding linear index in the destination tensor
-        dest_index = map_index(I.I, dest_strides_perm)
+    dest_strides_perm = ntuple(i->IT(dest_strides[findfirst(==(i), perm)]), N)
+    size_src = IT.(size(src))
+    LEN = IT(length(src))
```

Comment:
I know I'm nitpicking, but any reason to capitalize LEN? Resembles a type like that.

Reply:
Well, this is a legacy variable before we can store the length of arrays, it is not used anymore after switching to `@linearidx`, just removed it.